File: packages/react-native-autoplay/android/src/main/java/com/margelo/nitro/swe/iternio/reactnativeautoplay/HybridVoice.kt
package com.margelo.nitro.swe.iternio.reactnativeautoplay

import android.content.pm.PackageManager
import android.os.Build
import androidx.core.content.ContextCompat
import com.facebook.react.modules.core.PermissionAwareActivity
import com.facebook.react.modules.core.PermissionListener
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlin.coroutines.resume

class HybridVoice : HybridVoiceSpec() {
    @Volatile
    private var voiceInputManager: VoiceInputManager? = null

    override fun hasVoiceInputPermission(): Boolean {
        return VoiceInputManager.hasVoiceInputPermission()
    }

    override fun requestVoiceInputPermission(): Promise<Boolean> {
        return Promise.async {
            if (hasVoiceInputPermission()) {
                return@async true
            }

            val carContext = AndroidAutoSession.getRootContext()

            if (carContext != null) {
                suspendCancellableCoroutine { cont ->
                    carContext.requestPermissions(
                        listOf(android.Manifest.permission.RECORD_AUDIO)
                    ) { approved, _ ->
                        cont.resume(approved.contains(android.Manifest.permission.RECORD_AUDIO))
                    }
                }
            } else {
                val context = NitroModules.applicationContext ?: return@async false
                val activity =
                    context.currentActivity as? PermissionAwareActivity ?: return@async false
                val code = (Math.random() * 10000).toInt()

                suspendCancellableCoroutine { cont ->
                    activity.requestPermissions(
                        arrayOf(android.Manifest.permission.RECORD_AUDIO),
                        code,
                        PermissionListener { requestCode, _, grantResults ->
                            if (requestCode != code) {
                                return@PermissionListener false
                            }
                            cont.resume(
                                grantResults.isNotEmpty() &&
                                        grantResults.first() == PackageManager.PERMISSION_GRANTED
                            )
                            true
                        }
                    )
                }
            }
        }
    }

    override fun startVoiceInput(
        silenceThresholdMs: Double?,
        maxDurationMs: Double?,
        listeningText: String?,
        listeningImage: Variant_GlyphImage_AssetImage_RemoteImage?,
        listeningImageRepeats: Boolean?,
        preferSpeechToText: Boolean?,
        onChunk: ((chunk: VoiceInputChunk) -> Unit)?,
        language: String?,
        startSoundUri: String?,
        endSoundUri: String?,
        encoding: VoiceAudioEncoding?
    ): Promise<VoiceInputResult> {
        return Promise.async {
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
                throw UnsupportedOperationException("startVoiceInput requires at least API level ${Build.VERSION_CODES.O}")
            }

            val manager = VoiceInputManager(AndroidAutoSession.getRootContext())
            voiceInputManager = manager

            try {
                manager.start(
                    silenceThresholdMs = silenceThresholdMs?.toLong() ?: 1_500L,
                    maxDurationMs = maxDurationMs?.toLong() ?: 10_000L,
                    preferSpeechToText = preferSpeechToText ?: false,
                    onChunk = onChunk,
                    language = language,
                    startSoundUri = startSoundUri,
                    endSoundUri = endSoundUri,
                    encoding = encoding ?: VoiceAudioEncoding.LINEAR16,
                )
            } finally {
                voiceInputManager = null
                manager.dispose()
            }
        }
    }

    override fun stopVoiceInput() {
        voiceInputManager?.stop()
    }
}

File: packages/react-native-autoplay/ios/hybrid/HybridVoice.swift
import AVFoundation
import NitroModules
import Speech

class HybridVoice: HybridVoiceSpec {
    private var voiceInputManager: VoiceInputManager?

    func hasVoiceInputPermission() throws -> Bool {
        let micGranted = AVAudioSession.sharedInstance().recordPermission == .granted
        let speechGranted = SFSpeechRecognizer.authorizationStatus() == .authorized
        return micGranted && speechGranted
    }

    func requestVoiceInputPermission() throws -> Promise<Bool> {
        return Promise.async {
            let micGranted = await withCheckedContinuation { cont in
                AVAudioSession.sharedInstance().requestRecordPermission { granted in
                    cont.resume(returning: granted)
                }
            }
            guard micGranted else { return false }

            return await withCheckedContinuation { cont in
                SFSpeechRecognizer.requestAuthorization { status in
                    cont.resume(returning: status == .authorized)
                }
            }
        }
    }

    func startVoiceInput(
        silenceThresholdMs: Double?,
        maxDurationMs: Double?,
        listeningText: String?,
        listeningImage: Variant_GlyphImage_AssetImage_RemoteImage?,
        listeningImageRepeats: Bool?,
        preferSpeechToText: Bool?,
        onChunk: ((_ chunk: VoiceInputChunk) -> Void)?,
        language: String?,
        startSoundUri: String?,
        endSoundUri: String?,
        encoding: VoiceAudioEncoding?
    ) throws -> Promise<VoiceInputResult> {
        return Promise.async {
            let interfaceController = try? await RootModule.withInterfaceController { $0 }

            let manager = VoiceInputManager()
            self.voiceInputManager = manager

            defer { self.voiceInputManager = nil }

            return try await manager.start(
                interfaceController: interfaceController,
                silenceThresholdMs: silenceThresholdMs ?? 1_500,
                maxDurationMs: maxDurationMs ?? 10_000,
                listeningText: listeningText ?? "Listening...",
                listeningImage: listeningImage,
                listeningImageRepeats: listeningImageRepeats,
                preferSpeechToText: preferSpeechToText ?? false,
                onChunk: onChunk,
                language: language,
                startSoundUri: startSoundUri,
                endSoundUri: endSoundUri,
                encoding: encoding ?? .linear16
            )
        }
    }

    func stopVoiceInput() throws {
        Task { @MainActor in
            let interfaceController = try? await RootModule.withInterfaceController { $0 }
            self.voiceInputManager?.stop(interfaceController: interfaceController)
        }
    }
}
